ICE: `expected lifetime parameter, but found another generic parameter`

<!--
[31mICE[0m: Rustc ./a.rs '' 'error: internal compiler error: compiler/rustc_middle/src/ty/generics.rs:258:18: expected lifetime parameter, but found another generic parameter', 'error: internal compiler error: compiler/rustc_middle/src/ty/generics.rs:258:18: expected lifetime parameter, but found another generic parameter'
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
pub trait Foo2 {
    fn bar<'a: 'a>(&'a mut self) -> impl Sized + use<'static>;
}

impl Foo2 for () {
    fn bar<'a: 'a>(&'a mut self) -> impl Sized + 'a {}
}


````

original:
````rust
//@ check-pass
//@ edition:2021

pub trait Foo {
    #[allow(async_fn_in_trait)]
    async fn bar<'a: 'a>(&'a mut self);
}

impl Foo for () {
    async fn bar<'a: 'a>(&'a mut self) {}
}

pub trait Foo2 {
    fn bar<'a: 'a>(&'a mut self) -> impl Sized + use<'static>;
}

impl Foo2 for () {
    fn bar<'a: 'a>(&'a mut self) -> impl Sized + 'a {}
}

fn main() {}

````

Version information
````
rustc 1.82.0-nightly (a7399ba69 2024-08-31)
binary: rustc
commit-hash: a7399ba69d37b019677a9c47fe89ceb8dd82db2d
commit-date: 2024-08-31
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error: `use<...>` precise capturing syntax is currently not allowed in return-position `impl Trait` in traits
 --> /tmp/icemaker_global_tempdir.xgv2FYp3RLcf/rustc_testrunner_tmpdir_reporting.fx0w6Esy9qlw/mvce.rs:2:50
  |
2 |     fn bar<'a: 'a>(&'a mut self) -> impl Sized + use<'static>;
  |                                                  ^^^^^^^^^^^^
  |
  = note: currently, return-position `impl Trait` in traits and trait implementations capture all lifetimes in scope

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.xgv2FYp3RLcf/rustc_testrunner_tmpdir_reporting.fx0w6Esy9qlw/mvce.rs:7:2
  |
7 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.xgv2FYp3RLcf/rustc_testrunner_tmpdir_reporting.fx0w6Esy9qlw/mvce.rs`

error: expected lifetime parameter in `use<...>` precise captures list, found `'static`
 --> /tmp/icemaker_global_tempdir.xgv2FYp3RLcf/rustc_testrunner_tmpdir_reporting.fx0w6Esy9qlw/mvce.rs:2:54
  |
2 |     fn bar<'a: 'a>(&'a mut self) -> impl Sized + use<'static>;
  |                                                      ^^^^^^^

error: `impl Trait` must mention the `Self` type of the trait in `use<...>`
 --> /tmp/icemaker_global_tempdir.xgv2FYp3RLcf/rustc_testrunner_tmpdir_reporting.fx0w6Esy9qlw/mvce.rs:2:37
  |
1 | pub trait Foo2 {
  | -------------- `Self` type parameter is implicitly captured by this `impl Trait`
2 |     fn bar<'a: 'a>(&'a mut self) -> impl Sized + use<'static>;
  |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: currently, all type parameters are required to be mentioned in the precise captures list

error: internal compiler error: compiler/rustc_middle/src/ty/generics.rs:258:18: expected lifetime parameter, but found another generic parameter

thread 'rustc' panicked at compiler/rustc_middle/src/ty/generics.rs:258:18:
Box<dyn Any>
stack backtrace:
   0:     0x79ccab5fed9a - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::ha8b6a03154ea19c3
   1:     0x79ccabe03157 - core::fmt::write::hb1fec16e33daff29
   2:     0x79ccacde7fd1 - std::io::Write::write_fmt::h890fd89091c43625
   3:     0x79ccab60146b - std::panicking::default_hook::{{closure}}::h8e59b6e48c542f1f
   4:     0x79ccab6010de - std::panicking::default_hook::he45ee90607527b0d
   5:     0x79ccaa75f349 - std[70086fc653812fb1]::panicking::update_hook::<alloc[5793336eebe11ba]::boxed::Box<rustc_driver_impl[bde591410b6ce234]::install_ice_hook::{closure#0}>>::{closure#0}
   6:     0x79ccab601d87 - std::panicking::rust_panic_with_hook::h57f9e7cd2d85b015
   7:     0x79ccaa799551 - std[70086fc653812fb1]::panicking::begin_panic::<rustc_errors[d71da989a5965fd6]::ExplicitBug>::{closure#0}
   8:     0x79ccaa78cc16 - std[70086fc653812fb1]::sys::backtrace::__rust_end_short_backtrace::<std[70086fc653812fb1]::panicking::begin_panic<rustc_errors[d71da989a5965fd6]::ExplicitBug>::{closure#0}, !>
   9:     0x79ccaa787f66 - std[70086fc653812fb1]::panicking::begin_panic::<rustc_errors[d71da989a5965fd6]::ExplicitBug>
  10:     0x79ccaa7a2771 - <rustc_errors[d71da989a5965fd6]::diagnostic::BugAbort as rustc_errors[d71da989a5965fd6]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  11:     0x79ccaad6b764 - rustc_middle[454e73a6dc67a815]::util::bug::opt_span_bug_fmt::<rustc_span[fcf79a5fd0f12046]::span_encoding::Span>::{closure#0}
  12:     0x79ccaad5131a - rustc_middle[454e73a6dc67a815]::ty::context::tls::with_opt::<rustc_middle[454e73a6dc67a815]::util::bug::opt_span_bug_fmt<rustc_span[fcf79a5fd0f12046]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  13:     0x79ccaad511cb - rustc_middle[454e73a6dc67a815]::ty::context::tls::with_context_opt::<rustc_middle[454e73a6dc67a815]::ty::context::tls::with_opt<rustc_middle[454e73a6dc67a815]::util::bug::opt_span_bug_fmt<rustc_span[fcf79a5fd0f12046]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  14:     0x79cca88b3b40 - rustc_middle[454e73a6dc67a815]::util::bug::bug_fmt
  15:     0x79ccad6c4b01 - <rustc_middle[454e73a6dc67a815]::ty::generics::Generics>::region_param.cold
  16:     0x79cca92b7418 - <rustc_middle[454e73a6dc67a815]::ty::region::Region>::opt_param_def_id
  17:     0x79ccaa8e6a0c - <rustc_hir_analysis[7408769f7e89f40d]::check::compare_impl_item::RemapHiddenTyRegions as rustc_type_ir[5ab55af5d1323a37]::fold::FallibleTypeFolder<rustc_middle[454e73a6dc67a815]::ty::context::TyCtxt>>::try_fold_region
  18:     0x79ccaa86d7cf - <rustc_middle[454e73a6dc67a815]::ty::generic_args::GenericArg as rustc_type_ir[5ab55af5d1323a37]::fold::TypeFoldable<rustc_middle[454e73a6dc67a815]::ty::context::TyCtxt>>::try_fold_with::<rustc_hir_analysis[7408769f7e89f40d]::check::compare_impl_item::RemapHiddenTyRegions>
  19:     0x79ccaa8e6325 - <rustc_hir_analysis[7408769f7e89f40d]::check::compare_impl_item::RemapHiddenTyRegions as rustc_type_ir[5ab55af5d1323a37]::fold::FallibleTypeFolder<rustc_middle[454e73a6dc67a815]::ty::context::TyCtxt>>::try_fold_ty
  20:     0x79ccaa92c8ee - rustc_hir_analysis[7408769f7e89f40d]::check::compare_impl_item::collect_return_position_impl_trait_in_trait_tys::{closure#0}
  21:     0x79ccab13bcf3 - rustc_query_impl[9d4e2ec78ebde536]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9d4e2ec78ebde536]::query_impl::collect_return_position_impl_trait_in_trait_tys::dynamic_query::{closure#2}::{closure#0}, rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 8usize]>>
  22:     0x79ccabfa9dee - rustc_query_system[d8e0298fbce79578]::query::plumbing::try_execute_query::<rustc_query_impl[9d4e2ec78ebde536]::DynamicConfig<rustc_query_system[d8e0298fbce79578]::query::caches::DefIdCache<rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[9d4e2ec78ebde536]::plumbing::QueryCtxt, false>
  23:     0x79ccab13f6ce - rustc_query_impl[9d4e2ec78ebde536]::query_impl::collect_return_position_impl_trait_in_trait_tys::get_query_non_incr::__rust_end_short_backtrace
  24:     0x79ccac3cd0c0 - rustc_middle[454e73a6dc67a815]::query::plumbing::query_get_at::<rustc_query_system[d8e0298fbce79578]::query::caches::DefIdCache<rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 8usize]>>>
  25:     0x79cca8f5af7e - rustc_hir_analysis[7408769f7e89f40d]::collect::type_of::type_of
  26:     0x79ccabfab1aa - rustc_query_impl[9d4e2ec78ebde536]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9d4e2ec78ebde536]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 8usize]>>
  27:     0x79ccabfa9dee - rustc_query_system[d8e0298fbce79578]::query::plumbing::try_execute_query::<rustc_query_impl[9d4e2ec78ebde536]::DynamicConfig<rustc_query_system[d8e0298fbce79578]::query::caches::DefIdCache<rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[9d4e2ec78ebde536]::plumbing::QueryCtxt, false>
  28:     0x79ccabfa99a7 - rustc_query_impl[9d4e2ec78ebde536]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  29:     0x79ccac3cd0c0 - rustc_middle[454e73a6dc67a815]::query::plumbing::query_get_at::<rustc_query_system[d8e0298fbce79578]::query::caches::DefIdCache<rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 8usize]>>>
  30:     0x79cca95946d0 - rustc_hir_analysis[7408769f7e89f40d]::check::wfcheck::check_well_formed
  31:     0x79ccac779bc7 - rustc_query_impl[9d4e2ec78ebde536]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9d4e2ec78ebde536]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 1usize]>>
  32:     0x79ccac779e3e - rustc_query_system[d8e0298fbce79578]::query::plumbing::try_execute_query::<rustc_query_impl[9d4e2ec78ebde536]::DynamicConfig<rustc_query_system[d8e0298fbce79578]::query::caches::VecCache<rustc_hir[efc26d3c0a3a24aa]::hir_id::OwnerId, rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[9d4e2ec78ebde536]::plumbing::QueryCtxt, false>
  33:     0x79ccac779ba2 - rustc_query_impl[9d4e2ec78ebde536]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  34:     0x79ccac77a97f - rustc_hir_analysis[7408769f7e89f40d]::check::wfcheck::check_mod_type_wf
  35:     0x79ccac77a79f - rustc_query_impl[9d4e2ec78ebde536]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9d4e2ec78ebde536]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 1usize]>>
  36:     0x79ccac778e7b - rustc_query_system[d8e0298fbce79578]::query::plumbing::try_execute_query::<rustc_query_impl[9d4e2ec78ebde536]::DynamicConfig<rustc_query_system[d8e0298fbce79578]::query::caches::DefaultCache<rustc_span[fcf79a5fd0f12046]::def_id::LocalModDefId, rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[9d4e2ec78ebde536]::plumbing::QueryCtxt, false>
  37:     0x79ccac778c2d - rustc_query_impl[9d4e2ec78ebde536]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  38:     0x79ccac0bb03b - rustc_hir_analysis[7408769f7e89f40d]::check_crate
  39:     0x79ccac5ed73f - rustc_interface[71f2dea45baddd28]::passes::run_required_analyses
  40:     0x79ccacc9a65e - rustc_interface[71f2dea45baddd28]::passes::analysis
  41:     0x79ccacc9a631 - rustc_query_impl[9d4e2ec78ebde536]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9d4e2ec78ebde536]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 1usize]>>
  42:     0x79ccacd79cae - rustc_query_system[d8e0298fbce79578]::query::plumbing::try_execute_query::<rustc_query_impl[9d4e2ec78ebde536]::DynamicConfig<rustc_query_system[d8e0298fbce79578]::query::caches::SingleCache<rustc_middle[454e73a6dc67a815]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[9d4e2ec78ebde536]::plumbing::QueryCtxt, false>
  43:     0x79ccacd79a0f - rustc_query_impl[9d4e2ec78ebde536]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  44:     0x79ccacba2dae - rustc_interface[71f2dea45baddd28]::interface::run_compiler::<core[d1a550dca8d22a5b]::result::Result<(), rustc_span[fcf79a5fd0f12046]::ErrorGuaranteed>, rustc_driver_impl[bde591410b6ce234]::run_compiler::{closure#0}>::{closure#1}
  45:     0x79ccacc555d0 - std[70086fc653812fb1]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[71f2dea45baddd28]::util::run_in_thread_with_globals<rustc_interface[71f2dea45baddd28]::util::run_in_thread_pool_with_globals<rustc_interface[71f2dea45baddd28]::interface::run_compiler<core[d1a550dca8d22a5b]::result::Result<(), rustc_span[fcf79a5fd0f12046]::ErrorGuaranteed>, rustc_driver_impl[bde591410b6ce234]::run_compiler::{closure#0}>::{closure#1}, core[d1a550dca8d22a5b]::result::Result<(), rustc_span[fcf79a5fd0f12046]::ErrorGuaranteed>>::{closure#0}, core[d1a550dca8d22a5b]::result::Result<(), rustc_span[fcf79a5fd0f12046]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d1a550dca8d22a5b]::result::Result<(), rustc_span[fcf79a5fd0f12046]::ErrorGuaranteed>>
  46:     0x79ccacc55c3a - <<std[70086fc653812fb1]::thread::Builder>::spawn_unchecked_<rustc_interface[71f2dea45baddd28]::util::run_in_thread_with_globals<rustc_interface[71f2dea45baddd28]::util::run_in_thread_pool_with_globals<rustc_interface[71f2dea45baddd28]::interface::run_compiler<core[d1a550dca8d22a5b]::result::Result<(), rustc_span[fcf79a5fd0f12046]::ErrorGuaranteed>, rustc_driver_impl[bde591410b6ce234]::run_compiler::{closure#0}>::{closure#1}, core[d1a550dca8d22a5b]::result::Result<(), rustc_span[fcf79a5fd0f12046]::ErrorGuaranteed>>::{closure#0}, core[d1a550dca8d22a5b]::result::Result<(), rustc_span[fcf79a5fd0f12046]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d1a550dca8d22a5b]::result::Result<(), rustc_span[fcf79a5fd0f12046]::ErrorGuaranteed>>::{closure#1} as core[d1a550dca8d22a5b]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  47:     0x79ccacc55fab - std::sys::pal::unix::thread::Thread::new::thread_start::h6f36a038c6277257
  48:     0x79ccae31839d - <unknown>
  49:     0x79ccae39d49c - <unknown>
  50:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.82.0-nightly (a7399ba69 2024-08-31) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [collect_return_position_impl_trait_in_trait_tys] comparing an impl and trait method signature, inferring any hidden `impl Trait` types in the process
#1 [type_of] computing type of `<impl at /tmp/icemaker_global_tempdir.xgv2FYp3RLcf/rustc_testrunner_tmpdir_reporting.fx0w6Esy9qlw/mvce.rs:5:1: 5:17>::{synthetic#0}`
end of query stack
error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [collect_return_position_impl_trait_in_trait_tys] comparing an impl and trait method signature, inferring any hidden `impl Trait` types in the process
#1 [type_of] computing type of `<impl at /tmp/icemaker_global_tempdir.xgv2FYp3RLcf/rustc_testrunner_tmpdir_reporting.fx0w6Esy9qlw/mvce.rs:5:1: 5:17>::{synthetic#0}`

-->

--- a/compiler/rustc_hir_analysis/src/check/compare_impl_item.rs
+++ b/compiler/rustc_hir_analysis/src/check/compare_impl_item.rs
@@ -726,7 +726,7 @@ pub(super) fn collect_return_position_impl_trait_in_trait_tys<'tcx>(
                     num_trait_args,
                     num_impl_args,
                     def_id,
-                    impl_def_id: impl_m.container_id(tcx),
+                    impl_m_def_id: impl_m.def_id,
                     ty,
                     return_span,
                 }) {
@@ -844,12 +844,18 @@ where
 
 struct RemapHiddenTyRegions<'tcx> {
     tcx: TyCtxt<'tcx>,
+    /// Map from early/late params of the impl to identity regions of the RPITIT (GAT)
+    /// in the trait.
     map: FxIndexMap<ty::Region<'tcx>, ty::Region<'tcx>>,
     num_trait_args: usize,
     num_impl_args: usize,
+    /// Def id of the RPITIT (GAT) in the *trait*.
     def_id: DefId,
-    impl_def_id: DefId,
+    /// Def id of the impl method which owns the opaque hidden type we're remapping.
+    impl_m_def_id: DefId,
+    /// The hidden type we're remapping. Useful for diagnostics.
     ty: Ty<'tcx>,
+    /// Span of the return type. Useful for diagnostics.
     return_span: Span,
 }
 
@@ -885,8 +891,7 @@ impl<'tcx> ty::FallibleTypeFolder<TyCtxt<'tcx>> for RemapHiddenTyRegions<'tcx> {
             ty::ReLateParam(_) => {}
             // Remap early-bound regions as long as they don't come from the `impl` itself,
             // in which case we don't really need to renumber them.
-            ty::ReEarlyParam(ebr)
-                if ebr.index >= self.tcx.generics_of(self.impl_def_id).count() as u32 => {}
+            ty::ReEarlyParam(ebr) if ebr.index as usize >= self.num_impl_args => {}
             _ => return Ok(region),
         }
 
@@ -899,7 +904,7 @@ impl<'tcx> ty::FallibleTypeFolder<TyCtxt<'tcx>> for RemapHiddenTyRegions<'tcx> {
                 );
             }
         } else {
-            let guar = match region.opt_param_def_id(self.tcx, self.tcx.parent(self.def_id)) {
+            let guar = match region.opt_param_def_id(self.tcx, self.impl_m_def_id) {
                 Some(def_id) => {
                     let return_span = if let ty::Alias(ty::Opaque, opaque_ty) = self.ty.kind() {
                         self.tcx.def_span(opaque_ty.def_id)

--- a/compiler/rustc_middle/src/ty/generics.rs
+++ b/compiler/rustc_middle/src/ty/generics.rs
@@ -255,7 +255,9 @@ impl<'tcx> Generics {
         let param = self.param_at(param.index as usize, tcx);
         match param.kind {
             GenericParamDefKind::Lifetime => param,
-            _ => bug!("expected lifetime parameter, but found another generic parameter"),
+            _ => {
+                bug!("expected lifetime parameter, but found another generic parameter: {param:#?}")
+            }
         }
     }
 
@@ -264,7 +266,7 @@ impl<'tcx> Generics {
         let param = self.param_at(param.index as usize, tcx);
         match param.kind {
             GenericParamDefKind::Type { .. } => param,
-            _ => bug!("expected type parameter, but found another generic parameter"),
+            _ => bug!("expected type parameter, but found another generic parameter: {param:#?}"),
         }
     }
 
@@ -273,7 +275,7 @@ impl<'tcx> Generics {
         let param = self.param_at(param.index as usize, tcx);
         match param.kind {
             GenericParamDefKind::Const { .. } => param,
-            _ => bug!("expected const parameter, but found another generic parameter"),
+            _ => bug!("expected const parameter, but found another generic parameter: {param:#?}"),
         }
     }
 

--- a/tests/crashes/129850.rs
+++ b/tests/crashes/129850.rs
@@ -1,9 +1,0 @@
-//@ known-bug: rust-lang/rust#129850
-
-pub trait Foo2 {
-    fn bar<'a: 'a>(&'a mut self) -> impl Sized + use<'static>;
-}
-
-impl Foo2 for () {
-    fn bar<'a: 'a>(&'a mut self) -> impl Sized + 'a {}
-}

--- a/tests/ui/impl-trait/precise-capturing/rpitit-captures-more-method-lifetimes.rs
+++ b/tests/ui/impl-trait/precise-capturing/rpitit-captures-more-method-lifetimes.rs
@@ -1,0 +1,16 @@
+// Make sure we don't ICE when an RPITIT captures more method args than the
+// trait definition, which is not allowed. Due to the default lifetime capture
+// rules of RPITITs, this is only doable if we use precise capturing.
+
+pub trait Foo {
+    fn bar<'tr: 'tr>(&'tr mut self) -> impl Sized + use<Self>;
+    //~^ ERROR `use<...>` precise capturing syntax is currently not allowed in return-position `impl Trait` in traits
+}
+
+impl Foo for () {
+    fn bar<'im: 'im>(&'im mut self) -> impl Sized + 'im {}
+    //~^ ERROR return type captures more lifetimes than trait definition
+    //~| WARN impl trait in impl method signature does not match trait method signature
+}
+
+fn main() {}

--- a/tests/ui/impl-trait/precise-capturing/rpitit-captures-more-method-lifetimes.stderr
+++ b/tests/ui/impl-trait/precise-capturing/rpitit-captures-more-method-lifetimes.stderr
@@ -1,0 +1,42 @@
+error: `use<...>` precise capturing syntax is currently not allowed in return-position `impl Trait` in traits
+  --> $DIR/rpitit-captures-more-method-lifetimes.rs:6:53
+   |
+LL |     fn bar<'tr: 'tr>(&'tr mut self) -> impl Sized + use<Self>;
+   |                                                     ^^^^^^^^^
+   |
+   = note: currently, return-position `impl Trait` in traits and trait implementations capture all lifetimes in scope
+
+error: return type captures more lifetimes than trait definition
+  --> $DIR/rpitit-captures-more-method-lifetimes.rs:11:40
+   |
+LL |     fn bar<'im: 'im>(&'im mut self) -> impl Sized + 'im {}
+   |            ---                         ^^^^^^^^^^^^^^^^
+   |            |
+   |            this lifetime was captured
+   |
+note: hidden type must only reference lifetimes captured by this impl trait
+  --> $DIR/rpitit-captures-more-method-lifetimes.rs:6:40
+   |
+LL |     fn bar<'tr: 'tr>(&'tr mut self) -> impl Sized + use<Self>;
+   |                                        ^^^^^^^^^^^^^^^^^^^^^^
+   = note: hidden type inferred to be `impl Sized + 'im`
+
+warning: impl trait in impl method signature does not match trait method signature
+  --> $DIR/rpitit-captures-more-method-lifetimes.rs:11:40
+   |
+LL |     fn bar<'tr: 'tr>(&'tr mut self) -> impl Sized + use<Self>;
+   |                                        ---------------------- return type from trait method defined here
+...
+LL |     fn bar<'im: 'im>(&'im mut self) -> impl Sized + 'im {}
+   |                                        ^^^^^^^^^^^^^^^^
+   |
+   = note: add `#[allow(refining_impl_trait)]` if it is intended for this to be part of the public API of this crate
+   = note: we are soliciting feedback, see issue #121718 <https://github.com/rust-lang/rust/issues/121718> for more information
+   = note: `#[warn(refining_impl_trait_reachable)]` on by default
+help: replace the return type so that it matches the trait
+   |
+LL |     fn bar<'im: 'im>(&'im mut self) -> impl Sized {}
+   |                                        ~~~~~~~~~~
+
+error: aborting due to 2 previous errors; 1 warning emitted
+

